map/chart_generator.cpp

```diff
@@ -0,0 +1,215 @@
+  };
+
+  static AGG_INLINE void blend_pix(unsigned op, TValueType * p, unsigned cr, unsigned cg,
+                                   unsigned cb, unsigned ca, unsigned cover)
+  {
+    using TBlendTable = agg::comp_op_table_rgba<Color, Order>;
+    if (p[Order::A])
+    {
+      TBlendTable::g_comp_op_func[op](p, (cr * ca + ScaleMask) >> ScaleShift,
+                                      (cg * ca + ScaleMask) >> ScaleShift,
+                                      (cb * ca + ScaleMask) >> ScaleShift, ca, cover);
+    }
+    else
+    {
+      TBlendTable::g_comp_op_func[op](p, cr, cg, cb, ca, cover);
+    }
+  }
+};
+}  // namespace
+
+void NormalizeChartData(deque<double> const & distanceDataM,
+                        feature::TAltitudes const & altitudeDataM, size_t resultPointCount,
+                        vector<double> & uniformAltitudeDataM)
+{
+  uniformAltitudeDataM.clear();
+  if (distanceDataM.empty() || resultPointCount == 0 ||
+      distanceDataM.size() != altitudeDataM.size())
+    return;
+
+  if (!is_sorted(distanceDataM.cbegin(), distanceDataM.cend()))
+  {
+    LOG(LERROR, ("Route segment distances are not sorted."));
+    return;
+  }
+
+  uniformAltitudeDataM.resize(resultPointCount);
+  double const routeLenM = distanceDataM.back();
+  double const stepLenM = routeLenM / static_cast<double>(resultPointCount - 1);
+
+  auto const calculateAltitude = [&](double distFormStartM) {
+    if (distFormStartM <= distanceDataM.front())
+      return static_cast<double>(altitudeDataM.front());
+    if (distFormStartM >= distanceDataM.back())
+      return static_cast<double>(altitudeDataM.back());
+
+    auto const lowerIt = lower_bound(distanceDataM.cbegin(), distanceDataM.cend(), distFormStartM);
+    if (lowerIt == distanceDataM.cbegin())
+      return static_cast<double>(altitudeDataM.front());
+    if (lowerIt == distanceDataM.cend())
+      return static_cast<double>(altitudeDataM.back());
+
+    size_t const nextPointIdx = distance(distanceDataM.cbegin(), lowerIt);
+    CHECK_LESS(0, nextPointIdx, ());
+    size_t const prevPointIdx = nextPointIdx - 1;
+
+    if (distanceDataM[prevPointIdx] == distanceDataM[nextPointIdx])
+      return static_cast<double>(altitudeDataM[prevPointIdx]);
+
+    double const k = (altitudeDataM[nextPointIdx] - altitudeDataM[prevPointIdx]) /
+                     (distanceDataM[nextPointIdx] - distanceDataM[prevPointIdx]);
+    return static_cast<double>(altitudeDataM[prevPointIdx]) +
+           k * (distFormStartM - distanceDataM[prevPointIdx]);
+  };
+
+  uniformAltitudeDataM.resize(resultPointCount);
+  for (size_t i = 0; i < resultPointCount; ++i)
+    uniformAltitudeDataM[i] = calculateAltitude(static_cast<double>(i) * stepLenM);
+}
+
+void GenerateYAxisChartData(size_t height, double minMetersPerPxl,
+                            vector<double> const & altitudeDataM, vector<double> & yAxisDataPxl)
+{
+  uint32_t constexpr kHeightIndentPxl = 2.0;
+  yAxisDataPxl.clear();
```

Comment:
I'd prefer to have an empty yAxisDataPxl in case of early return. (Before the resize.)